### PR TITLE
Semi-automatic self-test procedure

### PR DESCRIPTION
**Pull request by tomas-pecserke**
Adds UART commands trigerring semi-automatic self-test procedure.

The test procedure test:
- [x] Buttons
- [x] Thumbstick
- [x] D-Hat
- [ ] Rotary Wheel
- ~~Gyro~~

### CODE REVIEW

**Comment by tomas-pecserke** [comment]
@marcos-diaz I'd like some advice on how to approach testing rotary wheel and gyros

File: src/headers/self_test.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright (C) 2022, Input Labs Oy.
+
+#pragma once
+
+void self_test();
+
+void self_test_buttons();
+void self_test_thumbstick();
+void self_test_dhat();
```
Comment (marcos-diaz) [inline src/headers/self_test.h]: I would merge dhat into buttons

File: src/self_test.c [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+#include "profile.h"
+#include "uart.h"
+
+void self_test_button_press(const char *buttonName, Button* button) {
+    printf("Press button '%s': WAITING", buttonName);
+
+    do {
+        uart_listen_char_limited();
+        sleep_ms(1);
+        bus_i2c_io_update_cache();
```
Comment (marcos-diaz) [inline src/self_test.c]: It can possibly use the new `bus_i2c_io_read()` without the cache (needs rebasing)

File: src/self_test.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        thumbstick->report(thumbstick);
+    } while (!button->virtual_press);
+
+    printf("\rMove thumbstick %s: OK     \n", buttonName);
+}
+
+void self_test_thumbstick_do(Thumbstick* thumbstick) {
+    printf("Testing thumbstick\n");
+
+    self_test_thumbstick_direction("up", &thumbstick->up, thumbstick);
```
Comment (marcos-diaz) [inline src/self_test.c:84]: Sort `left,right,top,down` for consistency

File: src/uart.c [hunk 4]
```diff
@@ -17,6 +18,11 @@ void uart_listen_char() {
         printf("UART: Bootsel mode\n");
         reset_usb_boot(0, 0);
     }
+
+    if (limited) {
```
Comment (marcos-diaz) [inline src/uart.c:22]: Why the `limited` UART logic?

File: src/uart.c [hunk 5]
```diff
@@ -26,4 +32,36 @@ void uart_listen_char() {
         config_write_init();
         config_print();
     }
+    if (input == 'T') {
```
Comment (marcos-diaz) [inline src/uart.c:35]: For now I would keep only this one, and then adding
```
test:
	screen -S alpakka -X stuff T
```
to Makefile

**Comment by marcos-diaz** [comment]
@tomas-pecserke 
For the gyro I don't think a test in here would add much value.


For the rotary there is some added difficulty because the interrupt logic, but we can tackle that later.

File: src/uart.c (hunk 4 above)
Comment (tomas-pecserke) [reply src/uart.c:22]: So we can still reset and reboot to BOOTSEL when stuck on some step in self-test.
It allows these 2 actions even when waiting for input and allowing all UART commands could lead to unintended behavior.

File: src/self_test.c (hunk 2 above)
Comment (tomas-pecserke) [reply src/self_test.c]: Current implementation of `button.is_pressed()` uses cache internally, I'd like to use same logic as buttons do to test, if the buttons work.

Would you like me to extract that logic into a method I can feed I2C result to? That way we have one implementation and both test and regular code using the same.

Comment (marcos-diaz) [reply src/self_test.c]: Ah true, makes sense to update the cache then 👍

**Review by marcos-diaz — APPROVED** [review]

**Pull request merged**